File: rib/restapi/crud/operations.py
# Copyright 2023 Two Six Technologies
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

""" Operation database CRUD functions """

# Python Library Imports
from sqlalchemy.orm import Session
from typing import List, Optional

# Local Python Library Imports
from rib.restapi.models.operations import (
    DbOperation,
    DbOperationLogLine,
    OperationState,
)


def get_operations_count(db_session: Session, target: str = None) -> int:
    """Get the total number of operations in the database"""
    query = db_session.query(DbOperation)

    if target:
        query = query.filter(DbOperation.target == target)

    return query.count()


def get_operations(
    db_session: Session, offset: int = 0, limit: int = 20, target: str = None
) -> List[DbOperation]:
    """Get operations from the database"""
    query = db_session.query(DbOperation)

    if target:
        query = query.filter(DbOperation.target == target)

    return (
        query.order_by(DbOperation.postedTime.desc()).offset(offset).limit(limit).all()
    )


def get_dangling_operations(db_session: Session) -> List[DbOperation]:
    """Get all operations with a state of PENDING or RUNNING"""
    return (
        db_session.query(DbOperation)
        .order_by(DbOperation.postedTime.asc())
        .filter(
            (DbOperation.state == OperationState.PENDING)
            | (DbOperation.state == OperationState.RUNNING)
        )
        .all()
    )


def get_operation(db_session: Session, operation_id: int) -> Optional[DbOperation]:
    """Get the specified operation from the database"""
    return db_session.query(DbOperation).filter(DbOperation.id == operation_id).first()


def create_operation(db_session: Session, operation: DbOperation) -> DbOperation:
    """Add the given operation to the database"""
    db_session.add(operation)
    db_session.commit()
    db_session.refresh(operation)
    return operation


def delete_non_running_operations(db_session: Session):
    """Delete operations that are not in a running state"""
    db_session.query(DbOperation).filter(
        DbOperation.state != OperationState.RUNNING
    ).delete()
    db_session.commit()


def get_operation_logs_count(db_session: Session, operation_id: int) -> int:
    """Get the total number of log lines for the specified operation"""
    return (
        db_session.query(DbOperationLogLine)
        .filter(DbOperationLogLine.operation_id == operation_id)
        .count()
    )


def get_operation_logs(
    db_session: Session, operation_id: int, offset: int = 0, limit: int = 1000
) -> List[DbOperationLogLine]:
    """Get log lines for the specified operation"""
    return (
        db_session.query(DbOperationLogLine)
        .filter(DbOperationLogLine.operation_id == operation_id)
        .order_by(DbOperationLogLine.time.asc())
        .offset(offset)
        .limit(limit)
        .all()
    )


def create_operation_log(
    db_session: Session, line: DbOperationLogLine
) -> DbOperationLogLine:
    """Add the given operation log line to the database"""
    db_session.add(line)
    db_session.commit()
    db_session.refresh(line)
    return line

File: webapp/src/features/deployments/types/operations.ts
// Copyright 2023 Two Six Technologies
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { RangeConfig } from '@/features/range-config';

type BaseDeploymentCreateRequest = {
  name: string;
  race_core: string;
  network_manager_kit: string;
  comms_channels: string[];
  comms_kits: string[];
  android_app: string;
  linux_app: string;
  node_daemon: string;
  artifact_manager_kits: string[];
  android_client_image: string;
  linux_client_image: string;
  linux_server_image: string;
  range_config: RangeConfig;
  fetch_plugins_on_start: boolean;
  no_config_gen: boolean;
  disable_config_encryption: boolean;
  cache: string;
  race_log_level: string;
};

export type LocalDeploymentCreateRequest = BaseDeploymentCreateRequest & {
  enable_gpu: boolean;
};

export type ConfigGenerationParameters = {
  network_manager_custom_args: string;
  comms_custom_args: Record<string, string>;
  artifact_manager_custom_args: Record<string, string>;
  max_iterations: number;
  force: boolean;
  skip_config_tar: boolean;
  timeout: number;
};

export type NodeOperationRequest = {
  force: boolean;
  nodes: string[] | null;
  timeout: number;
};

export type StandUpLocalDeploymentRequest = NodeOperationRequest & {
  no_publish: boolean;
};

export type StandUpAwsDeploymentRequest = NodeOperationRequest & {
  no_publish: boolean;
};

export type BootstrapNodeRequest = {
  force: boolean;
  introducer: string;
  target: string;
  passphrase: string;
  architecture: 'x86_64' | 'arm64-v8a';
  timeout: number;
};
